collect/cache/cuckooSentCache.go

```diff
@@ -0,0 +1,179 @@
+	return 0
+}
+
+func (t *cuckooDroppedRecord) DescendantCount() uint {
+	return 0
+}
+
+func (t *cuckooDroppedRecord) Count(*types.Span) {
+}
+
+// Make sure it implements TraceSentRecord
+var _ TraceSentRecord = (*cuckooDroppedRecord)(nil)
+
+type cuckooSentCache struct {
+	kept    *lru.Cache
+	dropped *CuckooTraceChecker
+	cfg     config.SampleCacheConfig
+
+	// The done channel is used to decide when to terminate the monitor
+	// goroutine. When resizing the cache, we write to the channel, but
+	// when terminating the system, call Stop() to close the channel.
+	// Either one causes the goroutine to shut down, and in resizing
+	// we then restart the monitor.
+	done chan struct{}
+}
+
+// Make sure it implements TraceSentCache
+var _ TraceSentCache = (*cuckooSentCache)(nil)
+
+func NewCuckooSentCache(cfg config.SampleCacheConfig) (TraceSentCache, error) {
+	stc, err := lru.New(int(cfg.KeptSize))
+	if err != nil {
+		return nil, err
+	}
+	dropped := NewCuckooTraceChecker(cfg.DroppedSize)
+
+	cache := &cuckooSentCache{
+		kept:    stc,
+		dropped: dropped,
+		cfg:     cfg,
+		done:    make(chan struct{}),
+	}
+	// TODO: metrics for when this puppy gets cycled
+	go cache.monitor()
+	return cache, nil
+}
+
+// goroutine to monitor the cache and cycle the size check periodically
+func (c *cuckooSentCache) monitor() {
+	ticker := time.NewTicker(c.cfg.SizeCheckInterval)
+	for {
+		select {
+		case <-ticker.C:
+			c.dropped.Maintain()
+		case <-c.done:
+			return
+		}
+	}
+}
+
+// Stop halts the monitor goroutine
+func (c *cuckooSentCache) Stop() {
+	close(c.done)
+}
+
+func (c *cuckooSentCache) Record(trace *types.Trace, keep bool) {
+	if keep {
+		// record this decision in the sent record LRU for future spans
+		sentRecord := cuckooKeptRecord{
+			rate:      trace.SampleRate,
+			spanCount: trace.DescendantCount(),
+		}
+		c.kept.Add(trace.TraceID, &sentRecord)
+		return
+	}
+	// if we're not keeping it, save it in the dropped trace filter
+	c.dropped.Add(trace.TraceID)
+}
+
+func (c *cuckooSentCache) Check(span *types.Span) (TraceSentRecord, bool) {
+	// was it dropped?
+	if c.dropped.Check(span.TraceID) {
+		// we recognize it as dropped, so just say so; there's nothing else to do
+		return &cuckooDroppedRecord{}, false
+	}
+	// was it kept?
+	if sentRecord, found := c.kept.Get(span.TraceID); found {
+		if sr, ok := sentRecord.(*cuckooKeptRecord); ok {
+			// if we kept it, then this span being checked needs counting too
+			sr.Count(span)
+			return sr, true
+		}
+	}
+	// we have no memory of this place
+	return nil, false
+}
+
+func (c *cuckooSentCache) Resize(cfg config.SampleCacheConfig) error {
+	stc, err := lru.New(int(cfg.KeptSize))
+	if err != nil {
+		return err
+	}
+
+	// grab all the items in the current cache; if it's larger than
+	// what will fit in the new one, discard the oldest ones
+	// (we don't have to do anything with the ones we discard, this is
+	// the trace decisions cache).
+	keys := c.kept.Keys()
```

Comment:
Do we need a lock here to protect the current set of keys and prevent adding new ones during the resize?

Reply:
Fixed in latest